New FUSE API endpoint: /iex/historicalPrice.

jamie-munro opened the pull request:
# Design
New FUSE API endpoint: */iex/historicalPrice*
With two request parameters:
>*Symbol*: symbol to retrieve data about
>*Date*: Date to query (in format YYYYMMDD)
    
Example: 
>*/iex/historicalPrice?symbol=AAPL&date=20181214*

Retrieves data about AAPL (apple) from 2018-12-14

This consumes the IEX endpoint:
>*/stock/**{symbol}**/chart/date/**{date}**?chartByDay=true&token=DUMMY_TOKEN*

See: https://iexcloud.io/docs/api/#historical-prices 

# Test
New test defined to test the historicalPrice endpoint with valid parameters: 
![test code](https://user-images.githubusercontent.com/115724737/196479199-55facf99-2432-4140-ab5a-d09c9903b97c.PNG)

New tests defined to test the historicalPrice endpoint with invalid parameters: 
![invalid tests](https://user-images.githubusercontent.com/115724737/196479283-3bb4b0bc-8a6f-4592-8d49-0b02727044df.PNG)

Wiremock simulation of IEX historical-prices added:
![wiremock](https://user-images.githubusercontent.com/115724737/196479344-cd4a4ab6-4eb4-49be-bc25-22dbf6b28049.PNG)

All tests passing:
![tests](https://user-images.githubusercontent.com/115724737/196479386-35872cf6-e511-4b62-9b63-8dd7dbfe8bac.PNG)

jmunro-verify-finance committed:
New FUSE API endpoint: /iex/historicalPrice.
New test defined to test the historicalPrice endpoint with valid parameters.
New tests defined to test the historicalPrice endpoint with invalid parameters.
Wiremock simulation of IEX historical-prices added.
All tests passing.

GalateaFarhaad commented on src/main/java/org/galatea/starter/domain/IexHistoricalPrice.java:
```diff
@@ -0,0 +1,20 @@
+public class IexHistoricalPrice {
+
+  private String symbol;
+  private Date date;
```
Try using classes from the Java 8 datetime API instead of the older java.util.Date. The new classes are better in a lot of ways so it would be good to get a bit more familiar.

GalateaFarhaad commented on src/main/java/org/galatea/starter/service/IexService.java:
```diff
@@ -45,5 +47,20 @@ public List<IexLastTradedPrice> getLastTradedPriceForSymbols(final List<String>
+   * @return historical pricing data for the given symbol on the given date
+   */
+  public List<IexHistoricalPrice> getHistoricalPrice(final List<String> symbol, final List<String> date) {
+    if ((symbol.size() == 1) && (date.size() == 1)) {
```
Since we only expect one symbol and one date, it would be better to take in a single value for each rather than a List.

GalateaFarhaad commented on src/main/java/org/galatea/starter/entrypoint/IexRestController.java:
```diff
@@ -48,4 +49,18 @@ public List<IexLastTradedPrice> getLastTradedPrice(
+          MediaType.APPLICATION_JSON_VALUE})
+  public List<IexHistoricalPrice> getHistoricalPrice(
+          @RequestParam(value = "symbol") final List<String> symbol,
+          @RequestParam(value = "date") final List<String> date) {
```
There are certain expectations on the values we eventually pass to the IEX API, e.g. the provided date should be valid. Try adding some validations so that if there is invalid data sent in, we can fail the request early.

jamie-munro replied:
When replacing line 63 with:
```@RequestParam(value = "date") @DateTimeFormat(pattern = "yyyyMMdd") final List<LocalDate> dates)```

Spring performs the validation for me, and will return 400: bad request if an invalid date is entered. Is this best practice or would we be better off accepting the parameter as a string and then performing our own conversion/validations?

GalateaFarhaad replied:
As a standard way of doing things that's fine.

There could be scenarios in which we may want to perform our own conversions/validations - for example, if we have some complex conversion or validation logic ("accept this pattern in this case, accept this other pattern in this other case"). But generally you'd know if you had to do something like that.

GalateaFarhaad commented on src/main/java/org/galatea/starter/service/IexClient.java:
```diff
@@ -32,4 +35,13 @@ public interface IexClient {
+   * @param date the date which should be queried
+   * @return historical pricing data for the given symbol on the given date
+   */
+  @GetMapping("/stock/{symbol}/chart/date/{date}?chartByDay=true&token=${spring.rest.iexToken}")
```
The IEX API docs show multiple possible values for range, rather than just `/stock/{symbol}/chart/date/{date}`. For example, `/stock/twtr/chart/max`, `/stock/twtr/chart/1m`. Try supporting those as well.

GalateaFarhaad commented on src/main/java/org/galatea/starter/entrypoint/IexRestController.java:
```diff
@@ -48,4 +49,18 @@ public List<IexLastTradedPrice> getLastTradedPrice(
+          MediaType.APPLICATION_JSON_VALUE})
+  public List<IexHistoricalPrice> getHistoricalPrice(
+          @RequestParam(value = "symbol") final List<String> symbol,
+          @RequestParam(value = "date") final List<String> date) {
```
Right now for `date` we just pass the value through to the IEX API, but sometimes we end up needing to do some date-based logic, which is easier when using an actual datetime class rather than a String. Try using one of the Java 8 datetime options here as well.

jmunro-verify-finance committed:
Changed date representations to java.time.LocalDate
Added support for IEX historical range feature
Added validation to dates
Added validation to ranges
Removed unnecessary lists
Added tests for range feature, and for invalid dates/ranges
All tests passing
